Add cmdline tab-completion for setting string options

ychin opened the pull request:
Add tab-completion for setting string options on the cmdline using `:set=` (along with `:set+=` and `:set-=`).

The existing tab completion for setting options currently only works when nothing is typed yet, and it only fills in with the existing value, e.g. when the user does `:set diffopt=<Tab>` it will be completed to `set diffopt=internal,filler,closeoff` and nothing else. This isn't too useful as a user usually wants auto-complete to suggest all the possible values, such as 'iblank', or 'algorithm:patience'.

For set= and set+=, this adds a new optional callback function for each option that can be invoked when doing completion. This allows for each option to have control over how completion works. For example, in 'diffopt', it will suggest the default enumeration, but if `algorithm:` is selected, it will further suggest different algorithm types like 'meyers' and 'patience'. When using set=, the existing option value will be filled in as the first choice to preserve the existing behavior (this helps doing small edits to the option value). When using set+= this won't happen as it doesn't make sense to append an option by its own value in general.

For flag list options (e.g. 'mouse' and 'guioptions'), completion will take into account existing typed values (and in the case of set+=, the existing option value) to make sure it doesn't suggest duplicates.

For set-=, there is a new `ExpandSettingSubtract` function which will handle flag list and comma-separated options smartly, by only suggesting values that currently exist in the option so you won't try to subtract the wrong value.

Note that Vim has some existing code that adds special handling for 'filetype', 'syntax', and misc dir options like 'backupdir'. This change preserves them as they already work, instead of converting to the new callback API for each option.

ychin committed:
Add cmdline tab-completion for setting string options

Add tab-completion for setting string options on the cmdline using
`:set=` (along with `:set+=` and `:set-=`).

The existing tab completion for setting options currently only works
when nothing is typed yet, and it only fills in with the existing value,
e.g. when the user does `:set diffopt=<Tab>` it will be completed to
`set diffopt=internal,filler,closeoff` and nothing else. This isn't too
useful as a user usually wants auto-complete to suggest all the possible
values, such as 'iblank', or 'algorithm:patience'.

For set= and set+=, this adds a new optional callback function for each
option that can be invoked when doing completion. This allows for each
option to have control over how completion works. For example, in
'diffopt', it will suggest the default enumeration, but if `algorithm:`
is selected, it will further suggest different algorithm types like
'meyers' and 'patience'. When using set=, the existing option value will
be filled in as the first choice to preserve the existing behavior. When
using set+= this won't happen as it doesn't make sense.

For flag list options (e.g. 'mouse' and 'guioptions'), completion will
take into account existing typed values (and in the case of set+=, the
existing option value) to make sure it doesn't suggest duplicates.

For set-=, there is a new `ExpandSettingSubtract` function which will
handle flag list and comma-separated options smartly, by only suggesting
values that currently exist in the option.

Note that Vim has some existing code that adds special handling for
'filetype', 'syntax', and misc dir options like 'backupdir'. This change
preserves them as they already work, instead of converting to the new
callback API for each option.

yegappan commented on src/option.c:
```diff
@@ -7348,8 +7350,16 @@ set_context_in_set_cmd(
     if ((nextchar == '-' || nextchar == '+' || nextchar == '^') && p[1] == '=')
     {
+	if (nextchar == '-')
+	{
```
To be consistent the style followed for the next if statement, can you remove the braces for this if statement?

ychin replied:
I can fix up the misc code styles issues. I do think it will be nice if at some point Vim can add a proper code formatter/linter that we can just run instead of having to manually conform. It's always a fair bit of work for the initial setup and code cleanups though so not suggesting right now.

yegappan commented on src/option.c:
```diff
@@ -7418,34 +7453,57 @@ set_context_in_set_cmd(
+    // An option that is a list of single-character flags should always start
+    // at the end as we don't complete words.
+    if (flags & P_FLAGLIST)
+    {
```
Even though, I personally prefer always using curly braces for an if statement, the coding style used in the Vim source base is to not use the curly braces if there is only one statement in the if condition.

yegappan commented on src/option.c:
```diff
@@ -7739,33 +7826,204 @@ ExpandOldSetting(int *numMatches, char_u ***matches)
     return OK;
 }
 
+// Expansion handler for :set=/:set+= when the option has a custom expansion handler.
```
For function headers, the style is to use the three part "/* */" comments.

yegappan reviewed (commented):
This is very nice.  Thanks for working on this.  This will require quite a few new tests to make sure all the expansion cases are covered.

ychin committed:
Fix coding style issues

ychin committed:
Add completion handlers for simple enum and flag list string options

'whichwrap': This option is quite weird as it's a comma-separated flag
list (the commas are actually optional). Fix up the logic for set-= to
handle it properly, and remove the ',' from WW_ALL so auto-complete
works properly.

'concealcursor': This option is a flag list option but wasn't tagged
correctly. Make sure to tag it as a P_FLAGLIST to make sure it follows
the same logic as other similar options (e.g. 'mouse').

ychin committed:
Add more complicated options that require some piping of data

Some of these require keeping in sync with the parsing code, as the
expansion code is using a separate char* array.

List of options:
- breakindentopt
- clipboard
- complete
- cursorlineopt
  - If we change the parsing code to use `opt_strings_flags()` instead,
    we wouldn't have to duplicate the array. But it works for now.
- eventignore
- keyprotocol
  - Only the protocol (the second part after the colon) is matched. The
    first part (term name) is not matched, since there is no good weay
    to do so.
- lispoptions
- printoptions
- rightleftcmd
- spelloptions
- wildmode

Also, make the options that use ExpandGeneric not sort the results to
give us more control.

ychin committed:
Add expansion for 'highlight', and also protect OOM for utilities

ychin committed:
Add expansion for 'listchars' and 'fillchars'

ychin committed:
Add expansion for 'previewpopup', 'completepopup', 'wincolor'

ychin committed:
Add tests for options auto-complete

ychin committed:
Add documentation

ychin committed:
Fix build issues

- Fix optiondef to make sure it works on tiny / gui builds
- Fix misc warnings

ychin committed:
Fix memory leaks

Make sure to clean up the match array if we found no matches because the
caller doesn't clean up returned array if the number of matches is 0.

Also, fix a bug in immediately returning instead of just setting a ret
variable.

ychin committed:
Fix tests to pass in different platforms

Deal with Windows escaping slashes differently on cmdline

Also fix 'bg' could either be dark or light.

ychin committed:
Add more tests to improve test coverage

zeertzjq commented on src/testdir/test_options.vim:
```diff
@@ -360,13 +360,17 @@ func Test_set_completion()
+  call feedkeys(":set spellsuggest=best,file:test_options.v\<Tab>\<C-B>\"\<CR>", 'xt')
+  call assert_equal("\"set spellsuggest=best,file:test_options.vim", @:)
 
   " Expand value for 'key'
   set key=abcd
   call feedkeys(":set key=\<Tab>\<C-B>\"\<CR>", 'xt')
   call assert_equal('"set key=*****', @:)
+  call feedkeys(":set key-=\<Tab>\<C-B>\"\<CR>", 'xt')
+  call assert_equal('"set key-=*****', @:)
   set key=
```
This seems to also need `if has('+key')`. 'key' option doesn't work if `FEAT_CRYPT` isn't enabled.

chrisbra replied:
I would remove the whole completion for `set key` (I mean what sense does it make to autocomplete `*****`) and I also don't want to accidentally leak the crypt-key. ah this already seems to happen with the current vim. So probably fine. Thanks!

ychin replied:
I was actually going to suggest that in a separate PR! Since I was looking into cmdline completion, I think `key` needs some hardening.

I was going to propose removing completion for `set key`, and also remove `set-=` / `set+=` / `set^=`. Those have some security ramifications since they allow you to guess a substring of the key. We can just keep `set=`.

chrisbra replied:
thanks, yes makes sense!

zeertzjq commented on src/testdir/test_options.vim:
```diff
@@ -382,6 +386,278 @@ func Test_set_completion()
 endfunc
 
+" Test handling of expanding individual string option values
+func Test_set_completion_string_values()
```
Most of the `assert_equal()` calls in this function have wrong order of arguments. This will make error message confusing if test fails.

ychin replied:
Oh right… the first item is the expected value. I internalized the weirdness when I was debugging this but didn't correct it. I could fix it in a follow-up commit.

zeertzjq commented on src/screen.c:
```diff
@@ -4956,6 +4955,32 @@ set_listchars_option(win_T *wp, char_u *val, int apply)
+    return (char_u*)filltab[idx].name;
+}
+
+/*
+ * Function given to ExpandGeneric() to obtain possible arguments of the
+ * 'listchars' option.
+ */
+    char_u *
+get_listchars_name(expand_T *xp UNUSED, int idx)
+{
+    if (idx >= (int)(sizeof(lcstab) / sizeof(lcstab[0])))
+	return NULL;
+
+    return (char_u*)lcstab[idx].name;
+}
```
These don't include `multispace` and `leadmultispace`

zeertzjq replied:
Created PR #13225 to fix this.

zeertzjq commented on src/option.c:
```diff
@@ -1837,7 +1837,7 @@ stropt_get_newval(
     else
     {
-	++arg;	// jump to after the '=' or ':'
+	++arg;	// joption_value2stringump to after the '=' or ':'
```
Looks like an accidental mistake

ychin replied:
Yeah it was pointed out in https://github.com/vim/vim/commit/900894b09a95398dfc75599e9f0aa2ea25723384#r128808467 as well. I have a follow-up PR that will fix this.

zeertzjq commented on src/option.c:
```diff
@@ -7418,34 +7451,55 @@ set_context_in_set_cmd(
+		while (s > xp->xp_pattern && *(s - 1) == '\\')
+		    --s;
+		if ((*p == ' ' && (xp->xp_backslash == XP_BS_THREE && (p - s) < 3))
+			|| (*p == ',' && (flags & P_COMMA) && ((p - s) % 1) == 0)
```
This `(p - s) % 1` looks strange. Isn't any number module 1 always 0?

ychin replied:
You are absolutely right. I only moved this code around and added the `P_COLON` check below it but I didn't change the logic for handling commas here. As you said, the logic here is wrong. I didn't want the scope of PR to grow any more which is why I didn't fix it.

I think the correct logic here is actually `(p - s) >= 2`. Currently, if you have a file/folder of the name `foo,bar`, the correct way to specify it in an option is do so something like `:set backupdir=dir1,foo\\,bar,dir2` (note that this is equivalent to `:set backupdir=dir1,foo\\\,bar,dir2`). Note that you will see that in the test cases that I added, they actually check that `set-=` will correctly escape the commas to insert those "\\" when doing `set path-=<Tab>`.

Note: there's another cmd auto-complete bug (that's why I wanted to fix this in another PR instead of growing in scope). In the above example, if you do auto-complete as follows: `set backupdir=fo<Tab>`, if will complete to `set backupdir=foo,bar`, forgetting to insert the `\\` there. In general the option setting syntax is a little weird regarding commas, as I believe the syntax is incomplete and doesn't allow certain types of names to be entered (you cannot enter a folder/file that ends with "\", because that slash will end up escaping the comma instead).

tldr: Yes it's broken, but it's broken before my change and I could submit another PR to fix it.

ychin replied:
The commas will be fixed in #13303
